library/src/tests/diagnostics.rs

```diff
@@ -156,3 +156,117 @@ fn check_stop_counting_function_without_start() {
         &Value::Int(-1),
     );
 }
+
+#[test]
+fn check_start_counting_qubits_for_one_allocation() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            use q = Qubit();
+            StopCountingQubits()
+        }",
+        &Value::Int(1),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_for_tuple_allocation() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            use (q0, q1) = (Qubit(), Qubit());
+            StopCountingQubits()
+        }",
+        &Value::Int(2),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_for_array_allocation() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            use qs = Qubit[2];
+            StopCountingQubits()
+        }",
+        &Value::Int(2),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_after_allocation_gives_zero() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            use q = Qubit();
+            StartCountingQubits();
+            StopCountingQubits()
+        }",
+        &Value::Int(0),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_sees_same_qubit_as_single_count() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            {
+                use q = Qubit();
+            }
+            {
+                use q = Qubit();
+            }
+            StopCountingQubits()
+        }",
+        &Value::Int(1),
+    );
+}
+
```

Comment:
It would be helpful to include a test that would call an operation between starting and stopping counting qubits. The most common pattern of using AllowAtMostNQubits I can find is 

                within {
                    AllowAtMostNQubits(2*N, "You are not allowed to allocate extra qubits");
                } apply {
                    AssertOperationsEqualReferenced(N,
                                                    ArbitraryBitPattern_Oracle_Challenge(_, pattern),
                                                    ArbitraryBitPattern_Oracle_Challenge_Reference(_, pattern));
                }

So a test could replicate it with a comparison of some library operations using CheckOperationsAreEqual, and checking that 2N qubits are a result

Reply:
I added `check_counting_qubits_works_with_allocation_in_operation_calls` at the end of the file that shows how this pattern could work now.